ddl/ddl_api.go

```diff
@@ -1351,6 +1349,16 @@ func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *
 	if meta.GetPartitionInfo() == nil {
 		return errors.Trace(ErrPartitionMgmtOnNonpartitioned)
 	}
+	// We don't support add hash type partition now.
+	if t.Meta().Partition.Type == model.PartitionTypeHash {
+		return errors.Trace(ErrUnsupportedAddPartition)
+	}
+
+	if t.Meta().Partition.Type == model.PartitionTypeRange {
```

Comment:
You can check `if len(spec.PartDefinitions) == 0` in buildPartitionInfo. and you can merge the two check together, like `if t.Meta().Partition.Type == model.PartitionTypeRange && if len(spec.PartDefinitions) == 0 `